### PR TITLE
Address I2C lockup issue

### PR DESCRIPTION
**Pull request by ShakenCodes**
Fixes issue where I2C driver will not timeout (as timeout functions were Ok(()) noops.)
Ensures master_stop() is called when an error occurs.
Does not identify or address the I2C issue that exposed these weaknesses.

### CODE REVIEW

File: embassy-stm32/src/i2c/v2.rs [hunk 1]
```diff
@@ -382,35 +383,59 @@ impl<'d, T: Instance, TXDMA, RXDMA> I2c<'d, T, TXDMA, RXDMA> {
         // I2C start
         //
         // ST SAD+W
-        Self::master_write(
+        match Self::master_write(
```
Comment (Dirbaio) [inline embassy-stm32/src/i2c/v2.rs:386]: please write these with "early return" style. This avoids too much neested indentation and keeps the error handling close to the source errors.


```rust
if let Err(e) = Self::master_write(..) {
   ... handle it
   return Err(e)
}

**Comment by Dirbaio** [comment]
Sorry for the delay, we weren't sure what the best API for timeouts would be. 

Could you make the timeout configurable, instead of hardcoded to 100ms? It should be a `pub timeout: Option<Duration>` in `Config`. So even if the user enables the `time` feature, they can choose to not enable timeouts. 

Other than that, PR looks good so if the timeout is configurable it should be good to go. Thanks! :)

**Comment by Dirbaio** [comment]
Sorry for the delay, we weren't sure what the best API for timeouts would be. 

Could you make the timeout configurable, instead of hardcoded to 100ms? It should be a `pub timeout: Option<Duration>` in `Config`. So even if the user enables the `time` feature, they can choose to not enable timeouts. 

Other than that, PR looks good so if the timeout is configurable it should be good to go. Thanks! :)

**Comment by ShakenCodes** [comment]
Thank you for the feedback. I will make the requested change shortly. (I am on a small work emergency right now, so it might be a few days...)

**Comment by ShakenCodes** [comment]
I was looking at incorporating the requested improvement to the I2C driver and noticed a file in the directory named [timeout.rs](http://timeout.rs/) that seems to be doing what was intended all along. Have I just been using the wrong instance of the I2C? It looks like if you use TimeoutI2c::new() around I2c::new(), you get the timeout functionality I thought was missing.
If this is correct, I am happy to cancel/retract this PR.

**Comment by Dirbaio** [comment]
oh, oops. I hadn't noticed that either. Then yes, we shouldn't add timeouts to the main i2c.

The "do stop on error" still seems a change worth doing. Could you send a PR with just that?

**Comment by ShakenCodes** [comment]
This is unneeded due to existing TimeoutI2c feature.
Will open a separate PR for calling master_stop(0 consistently.

**Pull request closed without merging**